src/include/transaction/deferred_action_manager.h

```diff
@@ -0,0 +1,108 @@
+#pragma once
+#include <queue>
+#include <utility>
+#include <vector>
+#include "transaction/timestamp_manager.h"
+#include "transaction/transaction_defs.h"
+
+namespace terrier::transaction {
+/**
+ * The deferred action manager tracks deferred actions and provides a function to process them
+ */
+class DeferredActionManager {
+ public:
+  /**
+   * Constructs a new DeferredActionManager
+   * @param timestamp_manager source of timestamps in the system
+   */
+  explicit DeferredActionManager(TimestampManager *timestamp_manager) : timestamp_manager_(timestamp_manager) {}
+
+  /**
+   * Adds the action to a buffered list of deferred actions.  This action will
+   * be triggered no sooner than when the epoch (timestamp of oldest running
+   * transaction) is more recent than the time this function was called.
+   * @param a functional implementation of the action that is deferred. @see DeferredAction
+   */
+  timestamp_t RegisterDeferredAction(const DeferredAction &a) {
+    timestamp_t result = timestamp_manager_->CurrentTime();
```

Comment:
Grabbing the time stamp outside the latch below allows violations of your stated “sorted by time stamp” invariant for the queue (see below).  I don’t think this has any practical impacts at this time (we’re not reliant on the invariant), but we cannot guarantee a sorted queue with this ordering.